uxarray/grid/grid.py

```diff
@@ -147,25 +153,35 @@ def from_dataset(cls,
             ``xarray.Dataset`` containing unstructured grid coordinates and connectivity variables
         use_dual : bool, default=False
             When reading in MPAS formatted datasets, indicates whether to use the Dual Mesh
+        source_grid_spec: str, default=None
+            TODO
         """
         if not isinstance(dataset, xr.Dataset):
             raise ValueError("Input must be an xarray.Dataset")
 
         # determine grid/mesh specification
-        source_grid_spec = _parse_grid_type(dataset)
-
-        if source_grid_spec == "Exodus":
-            grid_ds, source_dims_dict = _read_exodus(dataset)
-        elif source_grid_spec == "Scrip":
-            grid_ds, source_dims_dict = _read_scrip(dataset)
-        elif source_grid_spec == "UGRID":
-            grid_ds, source_dims_dict = _read_ugrid(dataset)
-        elif source_grid_spec == "MPAS":
-            grid_ds, source_dims_dict = _read_mpas(dataset, use_dual=use_dual)
-        elif source_grid_spec == "Shapefile":
-            raise ValueError("Shapefiles not yet supported")
+
+        if source_grid_spec is None:
+            # parse to detect source grid spec
+
+            source_grid_spec = _parse_grid_type(dataset)
+            if source_grid_spec == "Exodus":
+                grid_ds, source_dims_dict = _read_exodus(dataset)
+            elif source_grid_spec == "Scrip":
+                grid_ds, source_dims_dict = _read_scrip(dataset)
+            elif source_grid_spec == "UGRID":
+                grid_ds, source_dims_dict = _read_ugrid(dataset)
+            elif source_grid_spec == "MPAS":
+                grid_ds, source_dims_dict = _read_mpas(dataset,
+                                                       use_dual=use_dual)
+            elif source_grid_spec == "Shapefile":
+                raise ValueError("Shapefiles not yet supported")
+            else:
+                raise ValueError("Unsupported Grid Format")
         else:
```

Comment:
What if the user-provided `source_grid_spec` is not the actual source grid specification, which we could know from parsing through it?

Reply:
This is a good point.

While the user is mostly expected to use our API `open_grid` when working with standalone grids, `Grid.from_dataset` could still be used by the user.

I could keep `source_grid_spec` "hidden" from the user by adding it as a kwarg and checking for it. This way the user won't be tempted to pass it in, but we would still have the power to pass it through for our use case. What do you think?

Reply:
Yes, I think that should work